src/Plugin/Commerce/PaymentGateway/Heartland.php

```diff
@@ -0,0 +1,391 @@
+                $form_state->setError($form['public_key'], $this->t('Your public key does not match the mode (@mode).', ['@mode' => $values['mode']]));
+            }
+
+            if (($mode == 'test' && $secret_env == 'prod') || ($mode == 'live' && $secret_env == 'cert')) {
+                $form_state->setError($form['secret_key'], $this->t('Your secret key does not match the mode (@mode).', ['@mode' => $values['mode']]));
+            }
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function submitConfigurationForm(array &$form, FormStateInterface $form_state)
+    {
+        parent::submitConfigurationForm($form, $form_state);
+  
+        if (!$form_state->getErrors()) {
+            $values = $form_state->getValue($form['#parents']);
+            $this->configuration['public_key'] = $values['public_key'];
+            $this->configuration['secret_key'] = $values['secret_key'];
+            // $this->configuration['store_card_data'] = $values['store_card_data'];
+        }
+    }
+  
+    /**
+     * {@inheritdoc}
+     */
+    public function createPayment(PaymentInterface $payment, $capture = true)
+    {
+        $this->assertPaymentState($payment, ['new']);
+        $payment_method = $payment->getPaymentMethod();
+        $this->assertPaymentMethod($payment_method);
+
+        // Perform the create payment request here, throw an exception if it fails.
+        // See \Drupal\commerce_payment\Exception for the available exceptions.
+        // Remember to take into account $capture when performing the request.
+
+        // Set some variables to use for our transaction
+        $amount = $payment->getAmount();
+        $number = $amount->getNumber();
+        $currency = $amount->getCurrencyCode();
+        $remote_id = $payment_method_token = $payment_method->getRemoteId();
+
+        // Heartland SDK - Prepare / Charge Credit Card
+        // https://developer.heartlandpaymentsystems.com/Documentation/credit-card-payments/#prepare-to-charge-a-credit-card
+        $card = new CreditCardData();
+        $card->token = $payment_method_token;
+
+        // Charge or Authorize depending on your Transaction Mode
+        // Commerce -> Configuration -> Orders -> Checkout Flows -> Edit -> Payment -> Transaction mode
+        try {
+            if ($capture) {
+                $response = $card->charge($number)
+                    ->withCurrency($currency)
+                    ->execute();
+            } else {
+                $response = $card->authorize($number)
+                    ->withCurrency($currency)
+                    ->execute();
+            }
+
+            $remote_id = $response->transactionId;           
+            $next_state = $capture ? 'completed' : 'authorization';    
+            
+            // Commerce plugin sets state and remote Id for next step
+            $payment->setState($next_state);
+            $payment->setRemoteId($remote_id);
```

Comment:
Is the transaction ID viewable by a store admin? If not, is there anywhere we can use to display this (like a notes section)?

Reply:
Yes, it's viewable for any order under the Payments tab.